### PR TITLE
o/snapstate, tests: allow mixing revision and channel on snap install

### PR DESCRIPTION
**Pull request by andrewphelpsj**
This removes the code that silently removes channels from actions that are sent to the store when the action also contains a channel.

We do continue to remove the channel when an action either contains specific validation sets, or if the action's revision comes from an enforced validation set.

### CODE REVIEW

**Review by alfonsosanchezbeato — COMMENTED** [review]
I think this is fine, but I'm curious now about what is the behavior when a validation set specifies a revision but the currently tracked channel for that snap has not ever had that revisions, what would happen when trying to enforce the assertion?

File: overlord/snapstate/target.go
Comment (andrewphelpsj) [inline overlord/snapstate/target.go:400]: @alfonsosanchezbeato that is why we clear out the channel here and below when we have enforced validation sets

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks

File: overlord/snapstate/target.go
Comment (alfonsosanchezbeato) [reply overlord/snapstate/target.go:400]: Discussed offline: if a validation set requires a revision for a snap it does not matter if it was never published in the tracked channel of an already installed snap, it will be the one installed anyway.

File: overlord/snapstate/target.go
Comment (pedronis) [inline overlord/snapstate/target.go:717]: nitpick: I wonder if this should be closer to the code that checks if si.Channel and revOpts.Channel match? or that check should be moved closer to this?

**Review by pedronis — COMMENTED** [review]
changes look good but this code:

https://github.com/snapcore/snapd/blob/master/store/store_action.go#L423

also needs changing or if we are waiting for the store to land all the bits, it needs a TODO at least?

we should also check whether there's more code like that in the store package

File: overlord/snapstate/target.go
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:717]: With the changes introduced to `resolveChannel`, in my refactor PR, this block will be going away/changing anyways.

**Review by pedronis — COMMENTED** [review]
lgtm, but let's see what happens with the spread tests

**Comment by andrewphelpsj** [comment]
@alfonsosanchezbeato this has had some changes since your approval.

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM

File: overlord/snapstate/target.go
Comment (andrewphelpsj) [reply overlord/snapstate/target.go]: Good catch, I spoke with Przemysław about this but didn't update this appropriately. Fixed now.

File: overlord/snapstate/target.go
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:190]: Should I do that now? I figured that it would better to get the refactor PR merged first.

File: overlord/snapstate/target.go
Comment (pedronis) [inline overlord/snapstate/target.go:316]: do we know though what happens when we pass just stable on a refresh? refreshes don't tend to use default tracks

File: overlord/snapstate/target.go
Comment (pedronis) [reply overlord/snapstate/target.go:190]: no, as long as we don't forget

File: overlord/snapstate/target.go
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:316]: That shouldn't be a concern, as we don't actually track just "stable". I've updated the comment here to clarify that a bit.

**Review by pedronis — APPROVED** [review]
thanks

**Pull request merged**
